ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/storage/flat/BonsaiArchiveFlatDbStrategy.java

```diff
@@ -0,0 +1,406 @@
+   */
+  @Override
+  public void putFlatAccountStorageValueByStorageSlotHash(
+      final SegmentedKeyValueStorageTransaction transaction,
+      final Hash accountHash,
+      final Hash slotHash,
+      final Bytes storage) {
+
+    // get natural key from account hash and slot key
+    byte[] naturalKey = calculateNaturalSlotKey(accountHash, slotHash);
+    // keyNearest, use MIN_BLOCK_SUFFIX in the absence of a block context:
+    byte[] keyNearest = calculateArchiveKeyWithMinSuffix(context, naturalKey);
+
+    transaction.put(ACCOUNT_STORAGE_STORAGE, keyNearest, storage.toArrayUnsafe());
+  }
+
+  /*
+   * Removes the storage value for the given account hash and storage slot key, using the world state root hash supplier, storage root supplier, and node loader.
+   */
+  @Override
+  public void removeFlatAccountStorageValueByStorageSlotHash(
+      final SegmentedKeyValueStorageTransaction transaction,
+      final Hash accountHash,
+      final Hash slotHash) {
+
+    // get natural key from account hash and slot key
+    byte[] naturalKey = calculateNaturalSlotKey(accountHash, slotHash);
+    // insert a key suffixed with block context, with 'deleted account' value
+    byte[] keySuffixed = calculateArchiveKeyWithMinSuffix(context, naturalKey);
+
+    transaction.put(ACCOUNT_STORAGE_STORAGE, keySuffixed, DELETED_STORAGE_VALUE);
+  }
+
+  public static byte[] calculateNaturalSlotKey(final Hash accountHash, final Hash slotHash) {
+    return Bytes.concatenate(accountHash, slotHash).toArrayUnsafe();
+  }
+
+  public static byte[] calculateArchiveKeyWithMinSuffix(
+      final BonsaiContext context, final byte[] naturalKey) {
+    return calculateArchiveKeyWithSuffix(context, naturalKey, MIN_BLOCK_SUFFIX);
+  }
+
+  public static byte[] calculateArchiveKeyNoContextMinSuffix(final byte[] naturalKey) {
+    return Arrays.concatenate(naturalKey, MIN_BLOCK_SUFFIX);
+  }
+
+  public static byte[] calculateArchiveKeyNoContextMaxSuffix(final byte[] naturalKey) {
+    return Arrays.concatenate(naturalKey, MAX_BLOCK_SUFFIX);
+  }
+
+  public static Bytes calculateArchiveKeyWithMaxSuffix(
+      final BonsaiContext context, final byte[] naturalKey) {
+    return Bytes.of(calculateArchiveKeyWithSuffix(context, naturalKey, MAX_BLOCK_SUFFIX));
+  }
+
+  // TODO JF: move this out of this class so can be used with ArchiveCodeStorageStrategy without
+  // being static
+  public static byte[] calculateArchiveKeyWithSuffix(
+      final BonsaiContext context, final byte[] naturalKey, final byte[] orElseSuffix) {
+    // TODO: this can be optimized, just for PoC now
+    return Arrays.concatenate(
+        naturalKey,
+        context
+            .getBlockHeader()
+            .map(BlockHeader::getNumber)
+            .map(Bytes::ofUnsignedLong)
+            .map(Bytes::toArrayUnsafe)
+            .orElseGet(
+                () -> {
+                  // TODO: remove or rate limit these warnings
```

Comment:
is it something to manage in another PR ?

Reply:
I found on first startup of a new archive node I saw a flurry of these warnings, and then they didn't happen again. I think for now I'd be happier leaving in the warnings without any rate limiting, and we can see if we get reports of them being too noisy. If it's OK I'd like to leave the TODO comment in place for now.